### ISSUE TITLE
[Bug] SKI calculates interpolation weights for each training iteration

### ISSUE THREAD

**Issue by ymohit** [cornellius-gp/gpytorch#1816]
# 🚀 Feature Request

The following line calls to the model which inturns call grid_interpolation_kernel for each training iteration which seems inefficient as interpolation weights won't change during the entire training. 

https://github.com/cornellius-gp/gpytorch/blob/master/gpytorch/models/exact_gp.py#L257

Precisely, the following computation is performed for each iteration of hyper-parameter update for SKI which is required only once.

https://github.com/cornellius-gp/gpytorch/blob/master/gpytorch/kernels/grid_interpolation_kernel.py#L129-L14

This can be easily avoided by caching interpolation weights.

**Comment by gpleiss** [comment]
I think this is a reasonable request. Would you be willing to make a PR?

**Comment by ymohit** [comment]
Thanks for your response @gpleiss. 

I'm happy to make a PR. But as of now, I'm not very clear what is the preferred way to accommodate this. Does keeping weights in cache and as the property of GridInterpolationKernel looks good?

**Comment by gpleiss** [comment]
Awesome! Here's what I'm thinking:
- When the kernel is in training mode (`self.training == True`), then cache the `inputs` to [_compute_grid method](https://github.com/cornellius-gp/gpytorch/blob/ade5db8df1f8e4acbc64ec1dd8809e58506827f1/gpytorch/kernels/grid_interpolation_kernel.py#L129). Also cache `interp_values` and `interp_indices`.
- For future forward calls in training mode, maybe check that the inputs equal the cached inputs (and raise a warning if this is not true). If they are equal, then return the cached values.
- When you exit training mode (by calling `self.eval()`), then clear the cache.

We have a `cached` decorator that makes these sorts of caching operations easy (see [VariationalStrategy](https://github.com/cornellius-gp/gpytorch/blob/ade5db8df1f8e4acbc64ec1dd8809e58506827f1/gpytorch/variational/variational_strategy.py#L76) for an example of how it's used. The nice thing about this decorator is that [cached values are cleared when you switch between training mode and evaluation mode](https://github.com/cornellius-gp/gpytorch/blob/ade5db8df1f8e4acbc64ec1dd8809e58506827f1/gpytorch/module.py#L295).
